File: morphir-dotnet/src/Morphir.CodeModel/IR/Name.fs
// Copyright 2024 FINOS
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//     http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
namespace Morphir.IR

module Name =
    open Morphir.SDK.Maybe
    open Morphir.SDK

    [<Struct>]
    type Name =
        | Name of Parts: string list

        static member inline FromList(words: string list) : Name = Name words

        static member PartsFromString(input: string) : string list =
            let wordPattern =
                Regex.fromString "([a-zA-Z][a-z]*|[0-9]+)" |> Maybe.withDefault Regex.never

            Regex.find wordPattern input
            |> List.map (fun me -> me.Match)
            |> List.map String.toLower

    let inline fromList (words: string list) : Name = Name words

    let fromString (input: string) : Name =
        input |> Name.PartsFromString |> fromList

    let inline toList (Name name) : List<string> = name

    let capitalize string : string =
        match String.uncons string with
        | Just(headChar, tailString) -> String.cons (Char.toUpper headChar) tailString
        | Nothing -> string

    let toTitleCase name =
        name |> toList |> List.map capitalize |> String.join ""

    let toCamelCase (name: Name) =
        match name |> toList with
        | [] -> System.String.Empty
        | head :: tail -> tail |> List.map capitalize |> List.cons head |> String.join ""

    let toHumanWords name : List<string> =
        let words = toList name

        let join abbrev =
            abbrev |> String.join "" |> String.toUpper

        let rec process' prefix abbrev suffix =
            match suffix with
            | [] ->
                if (List.isEmpty abbrev) then
                    prefix
                else
                    List.append prefix [ join abbrev ]
            | first :: rest ->
                if (String.length first = 1) then
                    process' prefix (List.append abbrev [ first ]) rest
                else
                    match abbrev with
                    | [] -> process' (List.append prefix [ first ]) [] rest
                    | _ -> process' (List.append prefix [ join abbrev; first ]) [] rest

        process' [] [] words

    let toSnakeCase name = name |> toHumanWords |> String.join "_"

open Name

type NameBuilder() =
    member inline _.Combine(Name existing, Name other) : Name = existing @ other |> Name.fromList
    member inline _.Delay(f: _ -> Name) = f ()
    member inline _.Return(nameStr: string) = Name.fromString nameStr
    member inline _.ReturnFrom(name: Name) = name

    member inline _.Yield(input: string) = Name.fromString input

File: morphir/ir/core/src/morphir/ir/Name.scala
package morphir.ir

/** Generated based on IR.Name
*/
object Name{

  type Name = morphir.sdk.List.List[morphir.sdk.String.String]
  
  def capitalize(
    string: morphir.sdk.String.String
  ): morphir.sdk.String.String =
    morphir.sdk.String.uncons(string) match {
      case morphir.sdk.Maybe.Just((headChar, tailString)) => 
        morphir.sdk.String.cons(morphir.sdk.Char.toUpper(headChar))(tailString)
      case morphir.sdk.Maybe.Nothing => 
        string
    }
  
  def fromList(
    words: morphir.sdk.List.List[morphir.sdk.String.String]
  ): morphir.ir.Name.Name =
    words
  
  def fromString(
    string: morphir.sdk.String.String
  ): morphir.ir.Name.Name = {
    val wordPattern: morphir.sdk.Regex.Regex = morphir.sdk.Maybe.withDefault(morphir.sdk.Regex.never)(morphir.sdk.Regex.fromString("""([a-zA-Z][a-z]*|[0-9]+)"""))
    
    morphir.ir.Name.fromList(morphir.sdk.List.map(morphir.sdk.String.toLower)(morphir.sdk.List.map(((x: morphir.sdk.Regex.Match) =>
      x._match))(morphir.sdk.Regex.find(wordPattern)(string))))
  }
  
  def toCamelCase(
    name: morphir.ir.Name.Name
  ): morphir.sdk.String.String =
    morphir.ir.Name.toList(name) match {
      case Nil => 
        """"""
      case head :: tail => 
        morphir.sdk.String.join("""""")(morphir.sdk.List.cons(head)(morphir.sdk.List.map(morphir.ir.Name.capitalize)(tail)))
    }
  
  def toHumanWords(
    name: morphir.ir.Name.Name
  ): morphir.sdk.List.List[morphir.sdk.String.String] = {
    val words: morphir.sdk.List.List[morphir.sdk.String.String] = morphir.ir.Name.toList(name)
    
    def join(
      abbrev: morphir.sdk.List.List[morphir.sdk.String.String]
    ): morphir.sdk.String.String =
      morphir.sdk.String.toUpper(morphir.sdk.String.join("""""")(abbrev))
    
    {
      def process(
        prefix: morphir.sdk.List.List[morphir.sdk.String.String]
      )(
        abbrev: morphir.sdk.List.List[morphir.sdk.String.String]
      )(
        suffix: morphir.sdk.List.List[morphir.sdk.String.String]
      ): morphir.sdk.List.List[morphir.sdk.String.String] =
        suffix match {
          case Nil => 
            if (morphir.sdk.List.isEmpty(abbrev)) {
              prefix
            } else {
              morphir.sdk.List.append(prefix)(morphir.sdk.List(join(abbrev)))
            }
          case first :: rest => 
            if (morphir.sdk.Basics.equal(morphir.sdk.String.length(first))(morphir.sdk.Basics.Int(1))) {
              process(prefix)(morphir.sdk.List.append(abbrev)(morphir.sdk.List(first)))(rest)
            } else {
              abbrev match {
                case Nil => 
                  process(morphir.sdk.List.append(prefix)(morphir.sdk.List(first)))(morphir.sdk.List(
                  
                  ))(rest)
                case _ => 
                  process(morphir.sdk.List.append(prefix)(morphir.sdk.List(
                    join(abbrev),
                    first
                  )))(morphir.sdk.List(
                  
                  ))(rest)
              }
            }
        }
      
      name match {
        case word :: Nil => 
          if (morphir.sdk.Basics.equal(morphir.sdk.String.length(word))(morphir.sdk.Basics.Int(1))) {
            name
          } else {
            process(morphir.sdk.List(
            
            ))(morphir.sdk.List(
            
            ))(words)
          }
        case _ => 
          process(morphir.sdk.List(
          
          ))(morphir.sdk.List(
          
          ))(words)
      }
    }
  }
  
  def toHumanWordsTitle(
    name: morphir.ir.Name.Name
  ): morphir.sdk.List.List[morphir.sdk.String.String] =
    morphir.ir.Name.toHumanWords(name) match {
      case firstWord :: rest => 
        morphir.sdk.List.cons(morphir.ir.Name.capitalize(firstWord))(rest)
      case Nil => 
        morphir.sdk.List(
        
        )
    }
  
  def toList(
    words: morphir.ir.Name.Name
  ): morphir.sdk.List.List[morphir.sdk.String.String] =
    words
  
  def toSnakeCase(
    name: morphir.ir.Name.Name
  ): morphir.sdk.String.String =
    morphir.sdk.String.join("""_""")(morphir.ir.Name.toHumanWords(name))
  
  def toTitleCase(
    name: morphir.ir.Name.Name
  ): morphir.sdk.String.String =
    morphir.sdk.String.join("""""")(morphir.sdk.List.map(morphir.ir.Name.capitalize)(morphir.ir.Name.toList(name)))

}